File: frontend/http/provider-preset.ts
import { httpClient } from './client';

export type SdkVendor = 'openai' | 'google' | 'claude';

// 提供商预设接口
export interface ProviderPreset {
  id: string;
  preset_id: string;
  display_name: string;
  description: string | null;
  provider_type: 'native' | 'aggregator';
  transport: 'http' | 'sdk';
  sdk_vendor: SdkVendor | null;
  base_url: string;
  models_path: string;
  messages_path: string | null;
  chat_completions_path: string;
  responses_path: string | null;
  supported_api_styles: ('openai' | 'responses' | 'claude')[] | null;
  retryable_status_codes: number[] | null;
  custom_headers: Record<string, string> | null;
  static_models: any[] | null;
  created_at: string;
  updated_at: string;
}

// 创建请求
export interface CreateProviderPresetRequest {
  preset_id: string;
  display_name: string;
  description?: string;
  provider_type?: 'native' | 'aggregator';
  transport?: 'http' | 'sdk';
  sdk_vendor?: SdkVendor;
  base_url: string;
  models_path?: string;
  messages_path?: string;
  chat_completions_path?: string;
  responses_path?: string;
  supported_api_styles?: ('openai' | 'responses' | 'claude')[];
  retryable_status_codes?: number[];
  custom_headers?: Record<string, string>;
  static_models?: any[];
}

// 更新请求
export interface UpdateProviderPresetRequest {
  display_name?: string;
  description?: string;
  provider_type?: 'native' | 'aggregator';
  transport?: 'http' | 'sdk';
  sdk_vendor?: SdkVendor;
  base_url?: string;
  models_path?: string;
  messages_path?: string;
  chat_completions_path?: string;
  responses_path?: string;
  supported_api_styles?: ('openai' | 'responses' | 'claude')[];
  retryable_status_codes?: number[];
  custom_headers?: Record<string, string>;
  static_models?: any[];
}

// 列表响应
export interface ProviderPresetListResponse {
  items: ProviderPreset[];
  total: number;
}

// 提供商预设服务
export const providerPresetService = {
  // 获取预设列表（所有用户可访问）
  getProviderPresets: async (): Promise<ProviderPresetListResponse> => {
    const response = await httpClient.get('/provider-presets');
    return response.data;
  },

  // 创建预设（仅管理员）
  createProviderPreset: async (
    data: CreateProviderPresetRequest
  ): Promise<ProviderPreset> => {
    const response = await httpClient.post('/admin/provider-presets', data);
    return response.data;
  },

  // 更新预设（仅管理员）
  updateProviderPreset: async (
    presetId: string,
    data: UpdateProviderPresetRequest
  ): Promise<ProviderPreset> => {
    const response = await httpClient.put(
      `/admin/provider-presets/${presetId}`,
      data
    );
    return response.data;
  },

  // 删除预设（仅管理员）
  deleteProviderPreset: async (presetId: string): Promise<void> => {
    await httpClient.delete(`/admin/provider-presets/${presetId}`);
  },
};

File: backend/app/models/provider_preset.py
from __future__ import annotations

from sqlalchemy import Column, String, Text, text
from sqlalchemy.orm import Mapped, relationship

from app.db.types import JSONBCompat

from .base import Base, TimestampMixin, UUIDPrimaryKeyMixin


class ProviderPreset(UUIDPrimaryKeyMixin, TimestampMixin, Base):
    """管理员维护的官方 Provider 预设配置。"""

    __tablename__ = "provider_presets"

    preset_id: Mapped[str] = Column(String(50), unique=True, nullable=False, index=True)
    display_name: Mapped[str] = Column(String(100), nullable=False)
    description: Mapped[str | None] = Column(Text, nullable=True)
    provider_type: Mapped[str] = Column(
        String(16), nullable=False, server_default=text("'native'"), default="native"
    )
    transport: Mapped[str] = Column(
        String(16), nullable=False, server_default=text("'http'"), default="http"
    )
    sdk_vendor: Mapped[str | None] = Column(
        String(32),
        nullable=True,
        doc="When transport='sdk', identifies which official SDK implementation to use (e.g. openai/google/claude/vertexai).",
    )
    base_url: Mapped[str] = Column(String(255), nullable=False)
    models_path: Mapped[str | None] = Column(String(100), nullable=True)
    chat_completions_path: Mapped[str | None] = Column(String(100), nullable=True)
    messages_path: Mapped[str | None] = Column(String(100), nullable=True)
    responses_path: Mapped[str | None] = Column(String(100), nullable=True)
    supported_api_styles = Column(JSONBCompat(), nullable=True)
    retryable_status_codes = Column(JSONBCompat(), nullable=True)
    custom_headers = Column(JSONBCompat(), nullable=True)
    static_models = Column(JSONBCompat(), nullable=True)
    metadata_json = Column("metadata", JSONBCompat(), nullable=True)

    providers: Mapped[list["Provider"]] = relationship(
        "Provider",
        back_populates="preset",
        foreign_keys="Provider.preset_uuid",
        cascade="save-update",
    )


__all__ = ["ProviderPreset"]
